README.md
# AI ML - Human Attributes Detection with Facial Feature Extraction

## Age, Emotion, Gender recognition with Facial feature extraction

This facial attribute extraction program detects facial coordinates using FaceNet model and uses MXNet facial attribute extraction model for extracting 40 types of facial attributes. This solution detects Emotion, Age and Gender along with facial attributes.

Deep Learning Models used for the library are,

- FaceNet model used for facial landmark recognition.
- A trained lightened moon Mxnet model used for facial attribute extraction.
- AgeNet pre-trained caffe model used for Age detection.
- GenderNet caffe model used for Gender detection.

## Table of contents

- [Getting started](#getting-started)
- [Features](#features)
- [Facial Attribute Classes](#facial-attribute-classes)
- [Emotion Detection Classes](#emotion-detection-classes)
- [Gender Detection Classes](#gender-detection-classes)
- [Age Detection Classes](#age-detection-classes)
- [Usage](#usage)
- [Results](#results)
- [Want to Contribute?](#want-to-contribute)
- [Need Help / Support?](#need-help)
- [Collection of Other Components](#collection-of-components)
- [Changelog](#changelog)
- [Credits](#credits)
- [License](#license)
- [Keywords](#Keywords)

## Getting started

Prerequisites for running the code are:

- Python == 3.6
- python-opencv == 4.2.0
- numpy == 1.18.5
- pandas == 0.24.2
- mxnet == 1.6.0
- python-dotenv == 0.14.0

We have tested our program in above version, however you can use it in other versions as well.

## Features

- Face detection using FaceNet model
- Detects facial attribute of a face in an image.
- Detects Emotions on the face.
- Predicts Gender of the detected face.
- Predicts Age of the detected face.

### Facial Attribute Classes

    ["5_o_Clock_Shadow","Arched_Eyebrows","Attractive","Bags_Under_Eyes","Bald", "Bangs","Big_Lips","Big_Nose",
    "Black_Hair","Blond_Hair","Blurry","Brown_Hair","Bushy_Eyebrows","Chubby","Double_Chin","Eyeglasses","Goatee",
    "Gray_Hair", "Heavy_Makeup","High_Cheekbones","Male","Mouth_Slightly_Open","Mustache","Narrow_Eyes","No_Beard",
    "Oval_Face","Pale_Skin","Pointy_Nose","Receding_Hairline","Rosy_Cheeks","Sideburns","Smiling","Straight_Hair",
    "Wavy_Hair","Wearing_Earrings","Wearing_Hat","Wearing_Lipstick","Wearing_Necklace","Wearing_Necktie","Young"] 
    
### Emotion Detection Classes

   - Happy
   - Neutral
   - Surprise
   - Angry
   - Fear
   - Sad
   - Disgust

### Gender Detection Classes

   - Male
   - Female

### Age Detection Classes

    ['(0-2)', '(4-6)', '(8-12)', '(15-20)', '(25-32)', '(38-43)', '(48-53)', '(60-100)']
    
## Usage

Inside the project's directory run:

```
python predict.py
```
You can find sample images in the Dataset folder and results can be seen on the terminal. Results directory contains images with detected faces.


### Results
#### Original Image

<img src="images/image1.jpg" width = "300" height = "225"/>
<img src="images/image2.jpg" width = "300" height = "225"/>

#### Detection Results

<img src="images/result1.png" />

## Want to Contribute?

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).

-----

## Need Help? 

We also provide a free, basic support for all users who want to use image processing techniques for their projects. In case you want to customize this image enhancement technique for your development needs, then feel free to contact our [AI/ML developers](https://www.weblineindia.com/ai-ml-dl-development.html).

-----

## Collection of Components

We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development](https://www.weblineindia.com/software-development-resources.html).

------

## Changelog

Detailed changes for each release are documented in [CHANGELOG.md](./CHANGELOG.md).

## Credits

- Refered mxnet-face for attribute extraction.  [mxnet-face](https://github.com/tornadomeet/mxnet-face).
- Refered fer2013/IMDB for emotional classification.  [fer2013/IMDB](https://github.com/oarriaga/face_classification).
- Refered AgeGender recognition.  [AgeGender](https://github.com/spmallick/learnopencv/tree/master/AgeGender).

## License

[MIT](LICENSE)

[mit]: https://github.com/miguelmota/is-valid-domain/blob/e48e90f3ecd55431bbdba950eea013c2072d2fac/LICENSE

## Keywords

 Mxnet_face, facial_attribute_extraction, Age_recognition, gender_recognition, emotion_recognition, caffemodel, fer2013



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [released]
- V 1.0.0

## [1.0.0] - 2020-07-14

### Added
- Added facenet for face detection for mxnet

### Removed
- Removed Dlib face detection


mxnet_moon/lightened_moon.py
import argparse,logging
import numpy as np
import mxnet as mx

logger = logging.getLogger()
logger.setLevel(logging.INFO)

def norm_stat(d):
    return mx.nd.norm(d)/np.sqrt(d.size)
mon = mx.mon.Monitor(10, norm_stat)

def group(data, num_r, num, kernel, stride, pad, layer, down_sampling='pool'):
    if num_r > 0:
        conv_r = mx.symbol.Convolution(data=data, num_filter=num_r, kernel=(1,1), name=('conv%s_r' % layer))
        slice_r = mx.symbol.SliceChannel(data=conv_r, num_outputs=2, name=('slice%s_r' % layer))
        act_r = mx.symbol.maximum(slice_r[0], slice_r[1])
        conv = mx.symbol.Convolution(data=act_r, kernel=kernel, stride=stride, pad=pad, num_filter=num, name=('conv%s' % layer))
    else:
        conv = mx.symbol.Convolution(data=data, kernel=kernel, stride=stride, pad=pad, num_filter=num, name=('conv%s' % layer))
    slice = mx.symbol.SliceChannel(data=conv, num_outputs=2, name=('slice%s' % layer))
    act = mx.symbol.maximum(slice[0], slice[1])
    if down_sampling == "pool":
        pool = mx.symbol.Pooling(data=act, pool_type="max", kernel=(2, 2), stride=(2,2), name=('pool%s' % layer))
    elif down_sampling == "conv":
        pool = mx.symbol.Convolution(data=act, kernel=(3,3), stride=(2,2), pad=(1,1), num_filter=num/2, name=('pool%s' % layer))
    return pool

def before_flatten(use_fuse=True):
    data = mx.symbol.Variable(name="data")
    pool1 = group(data, 0, 96, (5,5), (1,1), (2,2), str(1))
    pool2 = group(pool1, 96, 192, (3,3), (1,1), (1,1), str(2))
    pool3 = group(pool2, 192, 384, (3,3), (1,1), (1,1), str(3))
    pool4 = group(pool3, 384, 256, (3,3), (1,1), (1,1), str(4))
    pool5 = group(pool4, 256, 256, (3,3), (1,1), (1,1), str(5))
    if use_fuse:
        # # down-sampling pool2
        pool2_fuse_1 = mx.symbol.Convolution(data=pool2, kernel=(3,3), stride=(2,2), pad=(1,1), num_filter=384, name="pool2_fuse_1")
        pool2_fuse_1_slice = mx.symbol.SliceChannel(data=pool2_fuse_1, num_outputs=2, name="pool2_fuse_1_slice")
        pool2_fuse_1_act = mx.symbol.maximum(pool2_fuse_1_slice[0], pool2_fuse_1_slice[1])
        pool2_fuse_2 = mx.symbol.Convolution(data=pool2_fuse_1_act, kernel=(3,3), stride=(2,2), pad=(1,1), num_filter=256, name="pool2_fuse_2")
        pool2_fuse_2_slice = mx.symbol.SliceChannel(data=pool2_fuse_2, num_outputs=2, name="pool2_fuse_2_slice")
        pool2_fuse_2_act = mx.symbol.maximum(pool2_fuse_2_slice[0], pool2_fuse_2_slice[1])
        pool2_fuse = mx.symbol.Pooling(data=pool2_fuse_2_act, pool_type="max", kernel=(2, 2), stride=(2,2), name='pool2_fuse')
        # down-sampling pool3
        pool3_fuse_1 = mx.symbol.Convolution(data=pool3, kernel=(3,3), stride=(2,2), pad=(1,1), num_filter=256, name="pool3_fuse_1")
        pool3_fuse_1_slice = mx.symbol.SliceChannel(data=pool3_fuse_1, num_outputs=2, name="pool3_fuse_1_slice")
        pool3_fuse_1_act = mx.symbol.maximum(pool3_fuse_1_slice[0], pool3_fuse_1_slice[1])
        pool3_fuse = mx.symbol.Pooling(data=pool3_fuse_1_act, pool_type="max", kernel=(2, 2), stride=(2,2), name='pool3_fuse')
        # down-sampling pool4
        pool4_fuse_1 = mx.symbol.Convolution(data=pool4, kernel=(1,1), stride=(1,1), pad=(0,0), num_filter=256, name="pool4_fuse_1")
        pool4_fuse_1_slice = mx.symbol.SliceChannel(data=pool4_fuse_1, num_outputs=2, name="pool4_fuse_1_slice")
        pool4_fuse_1_act = mx.symbol.maximum(pool4_fuse_1_slice[0], pool4_fuse_1_slice[1])
        pool4_fuse = mx.symbol.Pooling(data=pool4_fuse_1_act, pool_type="max", kernel=(2, 2), stride=(2,2), name='pool4_fuse')
        return mx.symbol.Concat(*[mx.symbol.Flatten(data=pool5), mx.symbol.Flatten(data = pool4_fuse + pool3_fuse + pool2_fuse)])
    else:
        return mx.symbol.Flatten(data=pool5)

def lightened_moon_feature(num_classes=40, use_fuse=True):
    flatten = before_flatten(use_fuse)
    fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=512, name="fc1")
    slice_fc1 = mx.symbol.SliceChannel(data=fc1, num_outputs=2, name="slice_fc1")
    fc1a = mx.symbol.maximum(slice_fc1[0], slice_fc1[1])
    drop1 = mx.symbol.Dropout(data=fc1a, p=0.5, name="drop1")
    fc2 = mx.symbol.FullyConnected(data=drop1, num_hidden=num_classes, name="fc2")
    return fc2

def lightened_moon(num_classes=40, use_fuse=True):
    fc2 = lightened_moon_feature(num_classes=num_classes, use_fuse=use_fuse)
    moon = mx.symbol.MoonOutput(data=fc2, src_dist_path='./src_dict.txt', name='Moon')
    return moon

def main():
    # symbol = lightened_moon(num_classes=40, use_fuse=False)
    symbol = lightened_moon(num_classes=40, use_fuse=True)

    devs = mx.cpu() if args.gpus is None else [mx.gpu(int(i)) for i in args.gpus.split(',')]
    epoch_size = args.num_examples / args.batch_size
    checkpoint = mx.callback.do_checkpoint(args.model_save_prefix)
    kv = mx.kvstore.create(args.kv_store)
    arg_params = None
    aux_params = None
    if args.retrain:
        _, arg_params, aux_params = mx.model.load_checkpoint(args.model_load_prefix, args.model_load_epoch)
    train = mx.io.ImageRecordIter(
        path_imglist = args.list_dir + 'celeba_train.lst',
        path_imgrec = args.data_dir + "celeba_train.rec",
        label_width = 40,
        data_name   = 'data',
        label_name  = 'Moon_label',
        data_shape  = (1, 128, 128),
        scale       = 1./255,
        batch_size  = args.batch_size,
        rand_crop   = True,
        rand_mirror = True,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)
    val = mx.io.ImageRecordIter(
        path_imglist = args.list_dir + 'celeba_val.lst',
        path_imgrec = args.data_dir + "celeba_val.rec",
        label_width = 40,
        data_name   = 'data',
        label_name  = 'Moon_label',
        batch_size  = args.batch_size,
        data_shape  = (1, 128, 128),
        scale       = 1./255,
        rand_crop   = True,
        rand_mirror = False,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)
    model = mx.model.FeedForward(
        ctx                = devs,
        symbol             = symbol,
        arg_params         = arg_params,
        aux_params         = aux_params,
        num_epoch          = 100,
        begin_epoch        = args.model_load_epoch,
        learning_rate      = args.lr,
        momentum           = 0.9,
        wd                 = 0.0005,
        lr_scheduler       = mx.lr_scheduler.FactorScheduler(step=4*max(int(epoch_size * 1), 1), factor=0.8, stop_factor_lr=1e-5),
        initializer        = mx.init.Xavier(factor_type="in", magnitude=2.34))
    model.fit(
        X                  = train,
        eval_data          = val,
        eval_metric        = ['multi_binary_acc'],
        kvstore            = kv,
        batch_end_callback = mx.callback.Speedometer(args.batch_size, 10),
        epoch_end_callback = checkpoint)
        # monitor            = mon)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="command for training lightened-moon")
    parser.add_argument('--gpus', type=str, help='the gpus will be used, e.g "0,1,2,3"')
    parser.add_argument('--data-dir', type=str, default='./', help='the input data directory')
    parser.add_argument('--list-dir', type=str, default='/home/adruser/AttGAN-Tensorflow/data/img_align_celeba/',
                        help='the directory which contain the training list file')
    parser.add_argument('--model-save-prefix', type=str, default='../model/lightened_moon/lightened_moon_fuse-symbol.json',
                        help='the prefix of the model to save')
    parser.add_argument('--lr', type=float, default=0.05, help='initialization learning reate')
    parser.add_argument('--batch-size', type=int, default=384, help='the batch size')
    parser.add_argument('--num-examples', type=int, default=159923, help='the number of training examples')
    parser.add_argument('--kv-store', type=str, default='local', help='the kvstore type')
    parser.add_argument('--model-load-prefix', type=str, default='../model/lightened_moon/lightened_moon_fuse', help='the prefix of the model to load')
    parser.add_argument('--model-load-epoch', type=int, default=0, help='load the model on an epoch using the model-load-prefix')
    parser.add_argument('--retrain', action='store_true', default=False, help='true means continue training')
    args = parser.parse_args()
    logging.info(args)
    main()

model/emotion/detectemotion.py
import os
import sys
import cv2
import numpy as np
from keras.models import load_model
from .utils.inference import draw_text
from .utils.datasets import get_labels
from .utils.inference import load_image
from .utils.inference import detect_faces
from .utils.inference import apply_offsets
from .utils.inference import draw_bounding_box
from .utils.preprocessor import preprocess_input
from .utils.inference import load_detection_model

# parameters for loading data and images
emotion_model_path = 'model/emotion/emotion.hdf5'
# loading models
emotion_classifier = load_model(emotion_model_path, compile=False)

class Emotional():
    def __init__(self):
        self.emotion_model_path = emotion_model_path

    """ Detect the emotion of the cropped images
        Emotions are 
        1. Happy
        2. Neutral
        3. Surprise
        4. Angry
        5. Fear
        6. Sad
        7. Disgust

    """
    def emotionalDet(self,imagepath,faceBox):
        
        emotion_model_path = self.emotion_model_path
        emotion_labels = get_labels('fer2013')
        gender_labels = get_labels('imdb')
        font = cv2.FONT_HERSHEY_SIMPLEX
        # hyper-parameters for bounding boxes shape
        emotion_offsets = (20, 40)
        emotion_offsets = (0, 0)
        
        # getting inputfor inference
        emotion_target_size = emotion_classifier.input_shape[1:3]
        # loading images
        rgb_image = load_image(imagepath, grayscale=False)
        gray_image = load_image(imagepath, grayscale=True)
        gray_image = np.squeeze(gray_image)
        gray_image = gray_image.astype('uint8')
        filename1 = os.path.basename(imagepath)
        faceEmotionList = []
        count = 0
        # for face_coordinates in faces:
        if faceBox:
            count += 1
            x1, x2, y1, y2 = apply_offsets(faceBox, emotion_offsets)
            gray_face = gray_image[y1:y2, x1:x2]
            try:
                gray_face = cv2.resize(gray_face, (emotion_target_size))
            except:
                print("Something went wrong")
            gray_face = preprocess_input(gray_face, True)
            gray_face = np.expand_dims(gray_face, 0)
            gray_face = np.expand_dims(gray_face, -1)
            emotion_label_arg = np.argmax(emotion_classifier.predict(gray_face))
            emotion_text = emotion_labels[emotion_label_arg]
            faceEmotionList.append(emotion_text)
        return faceEmotionList

model/emotion/utils/inference.py
import cv2
import matplotlib.pyplot as plt
import numpy as np
from keras.preprocessing import image

# def load_image(image_path, grayscale=False, color_mode="rgb", target_size=None):
#     # pil_image = image.load_img(image_path, grayscale, target_size)
#     pil_image = image.load_img(image_path, grayscale, target_size)
#     return image.img_to_array(pil_image)


def load_image(image_path, grayscale=False, target_size=None):
    color_mode = 'grayscale'
    if grayscale == False:
        color_mode = 'rgb'
    else:
        grayscale = False
    pil_image = image.load_img(image_path, grayscale, color_mode, target_size)
    return image.img_to_array(pil_image)

def load_detection_model(model_path):
    detection_model = cv2.CascadeClassifier(model_path)
    return detection_model

def detect_faces(detection_model, gray_image_array):
    return detection_model.detectMultiScale(gray_image_array, 1.3, 5)

def draw_bounding_box(face_coordinates, image_array, color):
    x, y, w, h = face_coordinates
    cv2.rectangle(image_array, (x, y), (x + w, y + h), color, 2)

def apply_offsets(face_coordinates, offsets):
    x, y, width, height = face_coordinates
    x_off, y_off = offsets
    return (x - x_off, x + width + x_off, y - y_off, y + height + y_off)

def draw_text(coordinates, image_array, text, color, x_offset=0, y_offset=0,
                                                font_scale=2, thickness=2):
    x, y = coordinates[:2]
    cv2.putText(image_array, text, (x + x_offset, y + y_offset),
                cv2.FONT_HERSHEY_SIMPLEX,
                font_scale, color, thickness, cv2.LINE_AA)

def get_colors(num_classes):
    colors = plt.cm.hsv(np.linspace(0, 1, num_classes)).tolist()
    colors = np.asarray(colors) * 255
    return colors



predict.py
''' Facial attribute extraction using mxnet and facenet '''
#--------------------------------
# Date : 10-07-2020
# Project : Facial Attribute Extraction
# Category : DeepLearning
# Company : weblineindia
# Department : AI/ML
#--------------------------------
import os
import cv2
import sys
import glob
import logging
import argparse
import numpy as np
import mxnet as mx
import pandas as pd
from pathlib import Path
from dotenv import load_dotenv
import model.emotion.detectemotion as ime
from mxnet_moon.lightened_moon import lightened_moon_feature

env_path = Path('.') / '.env'
load_dotenv(dotenv_path=env_path)

logger = logging.getLogger()
logger.setLevel(logging.INFO)
# import pdb

# Load path from .env
faceProto = os.getenv("FACEDETECTOR")
faceModel = os.getenv("FACEMODEL")
ageProto = os.getenv("AGEDETECTOR")
ageModel = os.getenv("AGEMODEL")
genderProto = os.getenv("GENDERDETECTOR")
genderModel = os.getenv("GENDERMODEL")
pathImg = os.getenv("IMGPATH")
APPROOT = os.getenv("APPROOT")

#Load face detection model
faceNet=cv2.dnn.readNet(faceModel,faceProto)
#Load age detection model
ageNet=cv2.dnn.readNet(ageModel,ageProto)
#Load gender detection model
genderNet=cv2.dnn.readNet(genderModel,genderProto)
#create instance for emotion detection
ed = ime.Emotional()


""" Detects face and extracts the coordinates"""
def getFaceBox(net, image, conf_threshold=0.7):
    image=image.copy()
    imageHeight=image.shape[0]
    imageWidth=image.shape[1]
    blob=cv2.dnn.blobFromImage(image, 1.0, (300, 300), [104, 117, 123], True, False)
    net.setInput(blob)
    detections=net.forward()
    faceBoxes=[]
    for i in range(detections.shape[2]):
        confidence=detections[0,0,i,2]
        if confidence>conf_threshold:
            x1=int(detections[0,0,i,3]*imageWidth)
            y1=int(detections[0,0,i,4]*imageHeight)
            x2=int(detections[0,0,i,5]*imageWidth)
            y2=int(detections[0,0,i,6]*imageHeight)
            faceBoxes.append([x1,y1,x2,y2])
            cv2.rectangle(image, (x1,y1), (x2,y2), (0,255,0), int(round(imageHeight/150)), 8)
    return image,faceBoxes



""" Detects age and gender """
def genderAge(image,faceBox):

    MODEL_MEAN_VALUES=(78.4263377603, 87.7689143744, 114.895847746)
    ageList=['(0-2)', '(4-6)', '(8-12)', '(15-20)', '(25-32)', '(38-43)', '(48-53)', '(60-100)']
    genderList=['Male','Female']
    
    padding=20
    face=image[max(0,faceBox[1]-padding):
        min(faceBox[3]+padding,image.shape[0]-1),max(0,faceBox[0]-padding)
        :min(faceBox[2]+padding, image.shape[1]-1)]
    blob=cv2.dnn.blobFromImage(face, 1.0, (227,227), MODEL_MEAN_VALUES, swapRB=False)

    # Predict the gender
    genderNet.setInput(blob)
    genderPreds=genderNet.forward()
    gender=genderList[genderPreds[0].argmax()]
    # Predict the age
    ageNet.setInput(blob)
    agePreds=ageNet.forward()
    age=ageList[agePreds[0].argmax()]
    # Return
    return gender,age




""" Function for gender detection,age detection and """            
def main():
    symbol = lightened_moon_feature(num_classes=40, use_fuse=True)
    devs = mx.cpu() if args.gpus is None else [mx.gpu(int(i)) for i in args.gpus.split(',')]
    _, arg_params, aux_params = mx.model.load_checkpoint('model/lightened_moon/lightened_moon_fuse', 82)

    ''' Loading Image from directory and writing attributes into .txt file'''
    img_dir = os.path.join(pathImg)
    if os.path.exists(img_dir):
        names = os.listdir(pathImg)      
        img_paths = [name for name in names]
        for imge in range(4005,4005+len(names)):

            imge = "{:06d}.jpg".format(imge)
            path = pathImg+str(imge)
            print("Image Path",path)
            # read img and drat face rect
            image = cv2.imread(path)
            img = cv2.imread(path, -1)
            
            resultImg,faceBoxes=getFaceBox(faceNet,image)
            if not faceBoxes:
                print("No face detected")

            # Loop throuth the coordinates
            for faceBox in faceBoxes:               

                print("#====Detected Age and Gender====#")

                gender,age = genderAge(image,faceBox)
                print('Gender',gender)
                print('Age',age)

                # Predict emotions in the image
                print("#====Detected Emotion===========#")
                
                emlist = ed.emotionalDet(path,faceBox)
                print(emlist)
        
                # Detect the facial attributes using mxnet
                # crop face area
                left = faceBox[0]
                width = faceBox[2] - faceBox[0]
                top = faceBox[1]
                height =  faceBox[3] - faceBox[1]
                right = faceBox[2]
                bottom = faceBox[3]
                gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
                pad = [0.25, 0.25, 0.25, 0.25] if args.pad is None else args.pad
                left = int(max(0, left - width*float(pad[0])))
                top = int(max(0, top - height*float(pad[1])))
                right = int(min(gray.shape[1], right + width*float(pad[2])))
                bottom = int(min(gray.shape[0], bottom + height*float(pad[3])))
                gray = gray[left:right, top:bottom]
                # resizing image and increasing the image size
                gray = cv2.resize(gray, (args.size, args.size))/255.0
                img = np.expand_dims(np.expand_dims(gray, axis=0), axis=0)
                # get image parameter from mxnet
                arg_params['data'] = mx.nd.array(img, devs)
                exector = symbol.bind(devs, arg_params ,args_grad=None, grad_req="null", aux_states=aux_params)
                exector.forward(is_train=False)
                exector.outputs[0].wait_to_read()
                output = exector.outputs[0].asnumpy()
                # 40 facial attributes
                text = ["5_o_Clock_Shadow","Arched_Eyebrows","Attractive","Bags_Under_Eyes","Bald", "Bangs","Big_Lips","Big_Nose",
                        "Black_Hair","Blond_Hair","Blurry","Brown_Hair","Bushy_Eyebrows","Chubby","Double_Chin","Eyeglasses","Goatee",
                        "Gray_Hair", "Heavy_Makeup","High_Cheekbones","Male","Mouth_Slightly_Open","Mustache","Narrow_Eyes","No_Beard",
                        "Oval_Face","Pale_Skin","Pointy_Nose","Receding_Hairline","Rosy_Cheeks","Sideburns","Smiling","Straight_Hair",
                        "Wavy_Hair","Wearing_Earrings","Wearing_Hat","Wearing_Lipstick","Wearing_Necklace","Wearing_Necktie","Young"]
                
                #Predict the results
                pred = np.ones(40)
                # create a list based on the attributes generated.
                attrDict = {}
                detectedAttributeList = []
                for i in range(40):
                    attr = text[i].rjust(20)
                    if output[0][i] < 0:
                        attrDict[attr] = 'No'
                    else:
                        attrDict[attr] = 'Yes'
                        detectedAttributeList.append(text[i])

                print("#====Detected Attributes========#")
                for attribute in detectedAttributeList:
                    print(attribute)                  
                # Write images into the results directory
                cv2.imwrite(APPROOT+'results/'+str(imge), resultImg) 



if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="predict the face attribution of one input image")
    parser.add_argument('--gpus', type=str, help='the gpus will be used, e.g "0,1,2,3"')
    parser.add_argument('--size', type=int, default=128,
                        help='the image size of lfw aligned image, only support squre size')
    parser.add_argument('--pad', type=float, nargs='+',
                                 help="pad (left,top,right,bottom) for face detection region")
    parser.add_argument('--model-load-prefix', dest = 'model_load_prefix', type=str, default='../model/lightened_moon/lightened_moon_fuse',
                        help='the prefix of the model to load')
    args = parser.parse_args()
       
    logging.info(args)
    main()
